userspace/libsinsp/metrics_collector.cpp

```diff
@@ -0,0 +1,798 @@
+
+	// Create the complete 3-lines text-based Prometheus exposition format https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md
+	std::string prometheus_text = "# HELP " + prometheus_metric_name_fully_qualified + " https://falco.org/docs/metrics/\n";
+	prometheus_text += "# TYPE " + prometheus_metric_name_fully_qualified + " " + std::string(metrics_metric_type_name_mappings_prometheus[metric.metric_type]) + "\n";
+	prometheus_text += prometheus_metric_name_fully_qualified;
+	prometheus_text += "{raw_name=\"" + std::string(metric.name) + "\"" ;
+	for (const auto& [key, value] : const_labels)
+	{
+		prometheus_text += "," + key + "=\"" + value + "\"" ;
+	}
+	prometheus_text += "} "; // white space at the end important!
+	switch (metric.type)
+	{
+	case METRIC_VALUE_TYPE_U32:
+		prometheus_text += std::to_string(metric.value.u32);
+		break;
+	case METRIC_VALUE_TYPE_S32:
+		prometheus_text += std::to_string(metric.value.s32);
+		break;
+	case METRIC_VALUE_TYPE_U64:
+		prometheus_text += std::to_string(metric.value.u64);
+		break;
+	case METRIC_VALUE_TYPE_S64:
+		prometheus_text += std::to_string(metric.value.s64);
+		break;
+	case METRIC_VALUE_TYPE_D:
+		prometheus_text += std::to_string(metric.value.d);
+		break;
+	case METRIC_VALUE_TYPE_F:
+		prometheus_text += std::to_string(metric.value.f);
+		break;
+	case METRIC_VALUE_TYPE_I:
+		prometheus_text += std::to_string(metric.value.i);
+		break;
+	default:
+		break;
+	}
+
+	prometheus_text += "\n";
+	return prometheus_text;
+}
+
+std::string prometheus_metrics_converter::convert_metric_to_text_prometheus(std::string_view metric_name, std::string_view prometheus_namespace, std::string_view prometheus_subsystem, const std::map<std::string, std::string>& const_labels) const
+{
+	// Create `prometheus_metric_name_fully_qualified`
+	std::string prometheus_metric_name_fully_qualified;
+	if (!prometheus_namespace.empty())
+	{
+		prometheus_metric_name_fully_qualified += std::string(prometheus_namespace) + "_";
+	}
+	if (!prometheus_subsystem.empty())
+	{
+		prometheus_metric_name_fully_qualified += std::string(prometheus_subsystem) + "_";
+	}
+	prometheus_metric_name_fully_qualified += std::string(metric_name) + "_info";
+
+	// Create the complete 3-lines text-based Prometheus exposition format https://github.com/prometheus/docs/blob/main/content/docs/instrumenting/exposition_formats.md
+	std::string prometheus_text = "# HELP " + prometheus_metric_name_fully_qualified + " https://falco.org/docs/metrics/\n";
+	prometheus_text += "# TYPE " + prometheus_metric_name_fully_qualified + " gauge\n";
+	prometheus_text += prometheus_metric_name_fully_qualified;
+	prometheus_text += "{raw_name=\"" + std::string(metric_name) + "\"" ;
+	for (const auto& [key, value] : const_labels)
+	{
+		prometheus_text += "," + key + "=\"" + value + "\"" ;
+	}
+	prometheus_text += "} 1\n";
+	return prometheus_text;
+}
```

Comment:
Using the conversion helper from above and factoring the text generation to reduce code duplication:
```suggestion
std::string prometheus_qualifier(std::string_view prometheus_namespace, std::string_view prometheus_subsystem)
{
	std::string qualifier;
	if (!prometheus_namespace.empty())
	{
		qualifier += std::string(prometheus_namespace) + "_";
	}
	if (!prometheus_subsystem.empty())
	{
		qualifier += std::string(prometheus_subsystem) + "_";
	}
	return qualifier;
}

std::string prometheus_exposition_text(std::string_view metric_qualified_name, std::string_view metric_name, std::string_view metric_type_name, std::string_view metric_value, const std::map<std::string, std::string>& const_labels)
{
	std::string fqn(metric_qualified_name);

	std::string prometheus_text = "# HELP " + fqn + " https://falco.org/docs/metrics/\n";
	prometheus_text += "# TYPE " + fqn + " " + std::string(metric_type_name) + "\n";
	prometheus_text += fqn;
	prometheus_text += "{raw_name=\"" + std::string(metric_name) + "\"" ;
	for (const auto& [key, value] : const_labels)
	{
		prometheus_text += "," + key + "=\"" + value + "\"" ;
	}
	prometheus_text += "} "; // white space at the end important!
	prometheus_text += std::string(metric_value);
	prometheus_text += "\n";
	return prometheus_text;
}

std::string prometheus_metrics_converter::convert_metric_to_text_prometheus(const metrics_v2& metric, std::string_view prometheus_namespace, std::string_view prometheus_subsystem, const std::map<std::string, std::string>& const_labels) const
{
	std::string prometheus_metric_name_fully_qualified = prometheus_qualifier(prometheus_namespace, prometheus_subsystem) + std::string(metric.name) + "_";
	// Remove native libs unit suffixes if applicable.
	RE2::GlobalReplace(&prometheus_metric_name_fully_qualified, s_libs_metrics_units_suffix_pre_prometheus_text_conversion, "");
	RE2::GlobalReplace(&prometheus_metric_name_fully_qualified, s_libs_metrics_banned_prometheus_naming_characters, "_");
	prometheus_metric_name_fully_qualified += std::string(metrics_unit_name_mappings_prometheus[metric.unit]);

	return prometheus_exposition_text(prometheus_metric_name_fully_qualified,
										metric.name,
									 	metrics_metric_type_name_mappings_prometheus[metric.metric_type],
										metric_value_to_text(metric),
										const_labels);
}

std::string prometheus_metrics_converter::convert_metric_to_text_prometheus(std::string_view metric_name, std::string_view prometheus_namespace, std::string_view prometheus_subsystem, const std::map<std::string, std::string>& const_labels) const
{
	return prometheus_exposition_text(prometheus_qualifier(prometheus_namespace, prometheus_subsystem) + std::string(metric_name) + "_info",
										metric_name,
    									"gauge",
										"1",
										const_labels);
}
```